lightning/src/ln/chan_utils.rs

```diff
@@ -691,58 +776,425 @@ impl HolderCommitmentTransaction {
+	/// Trust our pre-built transaction and derived transaction creation public keys.
+	///
+	/// This should only be used if you fully trust the builder of this object.  It should not
+	/// be used by an external signer - instead use the verify function.
+	pub fn trust(&self) -> TrustedCommitmentTransaction {
+		self.inner.trust()
+	}
+}
+
+/// A pre-built Bitcoin commitment transaction and its txid.
+#[derive(Clone)]
+pub struct BuiltCommitmentTransaction {
+	/// The commitment transaction
+	pub transaction: Transaction,
+	/// The txid for the commitment transaction.
+	///
+	/// This is provided as a performance optimization, instead of calling transaction.txid()
+	/// multiple times.
+	pub txid: Txid,
+}
+
+impl_writeable!(BuiltCommitmentTransaction, 0, { transaction, txid });
+
+impl BuiltCommitmentTransaction {
+	/// Get the SIGHASH_ALL sighash value of the transaction.
+	///
+	/// This can be used to verify a signature.
+	pub fn get_sighash_all(&self, funding_redeemscript: &Script, channel_value_satoshis: u64) -> Message {
+		let sighash = &bip143::SigHashCache::new(&self.transaction).signature_hash(0, funding_redeemscript, channel_value_satoshis, SigHashType::All)[..];
+		hash_to_message!(sighash)
+	}
+
+	/// Sign a transaction, either because we are counter-signing the counterparty's transaction or
+	/// because we are about to broadcast a holder transaction.
+	pub fn sign<T: secp256k1::Signing>(&self, funding_key: &SecretKey, funding_redeemscript: &Script, channel_value_satoshis: u64, secp_ctx: &Secp256k1<T>) -> Signature {
+		let sighash = self.get_sighash_all(funding_redeemscript, channel_value_satoshis);
+		secp_ctx.sign(&sighash, funding_key)
+	}
+}
+
+/// This class tracks the per-transaction information needed to build a commitment transaction and to
+/// actually build it and sign.  It is used for holder transactions that we sign only when needed
+/// and for transactions we sign for the counterparty.
+///
+/// This class can be used inside a signer implementation to generate a signature given the relevant
+/// secret key.
+#[derive(Clone)]
+pub struct CommitmentTransaction {
+	commitment_number: u64,
+	to_broadcaster_value_sat: u64,
+	to_countersignatory_value_sat: u64,
+	feerate_per_kw: u32,
+	htlcs: Vec<HTLCOutputInCommitment>,
+	// A cache of the parties' pubkeys required to construct the transaction, see doc for trust()
+	keys: TxCreationKeys,
+	// For access to the pre-built transaction, see doc for trust()
+	built: BuiltCommitmentTransaction,
+}
+
+impl PartialEq for CommitmentTransaction {
+	fn eq(&self, o: &Self) -> bool {
+		return self.commitment_number == o.commitment_number &&
+			self.to_broadcaster_value_sat == o.to_broadcaster_value_sat &&
+			self.to_countersignatory_value_sat == o.to_countersignatory_value_sat &&
+			self.feerate_per_kw == o.feerate_per_kw &&
+			self.htlcs == o.htlcs &&
+			self.keys == o.keys &&
+			self.check_built_transaction(o);
```

Comment:
It's here because we want to shortcut it away if any previous comparisons fail, because the debug assertion is only appropriate if all the non-derived fields are equal.  We could inline it by using a temp var for the non-derived field comparison result, but that seems more cluttered.  What do you think?

Reply:
Right, would require an if block or a temp var. I still think its worth it because `check_built_transaction` sounds a lot like `verify`-related stuff to me, and in general having highly related (ie the only caller) code 100 lines down just makes things harder to follow.

Reply:
Makes sense, inlined.